### PR TITLE
Improve the warning messages for the `#[diagnostic::on_unimplemented]`

### PR DESCRIPTION
**Pull request by weiznich**
This commit improves warnings emitted for malformed on unimplemented attributes by:

* Improving the span of the warnings
* Adding a label message to them
* Separating the messages for missing and unexpected options
* Adding a help message that says which options are supported

r? @compiler-errors 

I'm happy to work on further improvements, so feel free to make suggestions.

### CODE REVIEW

File: compiler/rustc_trait_selection/messages.ftl [hunk 1]
```diff
@@ -28,6 +28,11 @@ trait_selection_invalid_on_clause_in_rustc_on_unimplemented = invalid `on`-claus
     .label = invalid on-clause here
 
 trait_selection_malformed_on_unimplemented_attr = malformed `on_unimplemented` attribute
+    .help = Only `message`, `note` and `label` are allowed as options
```
Comment (Urgau) [inline compiler/rustc_trait_selection/messages.ftl:31]:
```suggestion
    .help = only `message`, `note` and `label` options are allowed
```

**Review by Urgau — COMMENTED** [review]
Diagnostics should not start with a Upper case.

There is some guidelines on the rustc-dev-guide https://rustc-dev-guide.rust-lang.org/diagnostics.html#suggestion-style-guide but unfortunatly not enough.

File: tests/ui/diagnostic_namespace/on_unimplemented/ignore_unsupported_options_and_continue_to_use_fallback.stderr [hunk 2]
```diff
@@ -22,8 +23,9 @@ LL | |     if(Self = ()),
 ...  |
 LL | |     note = "not used yet"
 LL | | )]
-   | |__^
+   | |__^ invalid option found here
```
Comment (compiler-errors) [inline tests/ui/diagnostic_namespace/on_unimplemented/ignore_unsupported_options_and_continue_to_use_fallback.stderr:26]: Should be able to point at the `if(Self = ())` here.

Comment (weiznich) [reply tests/ui/diagnostic_namespace/on_unimplemented/ignore_unsupported_options_and_continue_to_use_fallback.stderr]: I've pushed a "fix" to address this somewhat. The problem there is that `if(Self = ())` doesn't parse as `meta_item_list()`. I'm not sure why that's the case or whether that's the right behavior. The pushed "fix" just turns the attribute into something that can be parsed (`if(Self = "()")` in that case).

Comment (compiler-errors) [reply tests/ui/diagnostic_namespace/on_unimplemented/ignore_unsupported_options_and_continue_to_use_fallback.stderr]: oh, ok

**Pull request merged**